File: windows/YtRec.Core/DiskGuard.cs
namespace YtRec.Core;

public enum DiskState { Ok, Warn, Refuse }

/// <summary>Result of a pre-recording disk check. Carries the free-byte count for Warn/Refuse.</summary>
public readonly record struct DiskPreCheck(DiskState State, long FreeBytes)
{
    public static readonly DiskPreCheck Ok = new(DiskState.Ok, 0);
    public static DiskPreCheck Warn(long bytes) => new(DiskState.Warn, bytes);
    public static DiskPreCheck Refuse(long bytes) => new(DiskState.Refuse, bytes);
}

/// <summary>Disk protection (pure logic). Decimal GB to match Explorer/Finder. Mirrors mac DiskGuard.swift.</summary>
public static class DiskGuard
{
    public const double Gb = 1_000_000_000.0;

    public static DiskPreCheck PreCheck(long freeBytes, double warnGB = 15, double refuseGB = 8)
    {
        var free = freeBytes / Gb;
        if (free < refuseGB) return DiskPreCheck.Refuse(freeBytes);
        if (free < warnGB) return DiskPreCheck.Warn(freeBytes);
        return DiskPreCheck.Ok;
    }

    /// <summary>During recording: stop &amp; finalize if free space drops below the stop threshold.</summary>
    public static bool ShouldStopRecording(long freeBytes, double stopGB = 10) =>
        freeBytes / Gb < stopGB;

    /// <summary>Injectable free-bytes resolver: null (unknown) → long.MaxValue so we never falsely block.</summary>
    public static long FreeBytes(Func<long?> resolver) => resolver() ?? long.MaxValue;
}

File: mac/Sources/YTRec/Support/DiskGuard.swift
import Foundation

/// 磁碟保護（純邏輯，可測試）。閾值以十進位 GB 計，對齊 Finder 顯示。
enum DiskGuard {
    static let gb = 1_000_000_000.0

    /// 啟動側錄前的判定
    enum PreCheck: Equatable {
        case ok                 // 充足
        case warn(Int64)        // 偏低，警告後可續錄
        case refuse(Int64)      // 不足，拒絕啟動側錄（下載軌仍可跑）
    }

    static func preCheck(freeBytes: Int64, warnGB: Double = 15, refuseGB: Double = 8) -> PreCheck {
        let free = Double(freeBytes) / gb
        if free < refuseGB { return .refuse(freeBytes) }
        if free < warnGB { return .warn(freeBytes) }
        return .ok
    }

    /// 側錄中每隔一段時間檢查：低於門檻就收工保檔
    static func shouldStopRecording(freeBytes: Int64, stopGB: Double = 10) -> Bool {
        Double(freeBytes) / gb < stopGB
    }

    /// 可注入版（測試用）：resolver 回 nil＝查不到 → .max（不誤擋）。
    static func freeBytes(resolver: () -> Int64?) -> Int64 {
        resolver() ?? .max
    }

    /// 查詢某路徑所在磁碟的可用空間。查不到時回傳 .max（不誤擋）。
    static func freeBytes(at url: URL) -> Int64 {
        freeBytes {
            let keys: Set<URLResourceKey> = [.volumeAvailableCapacityForImportantUsageKey]
            if let v = try? url.resourceValues(forKeys: keys),
               let bytes = v.volumeAvailableCapacityForImportantUsage {
                return Int64(bytes)
            }
            return nil
        }
    }
}
